sourcegraph/sourcegraph
# Update browser extention installation detection logic on web

Closes https://github.com/sourcegraph/sourcegraph/issues/30842.


## Description
This PR changes the way how we detect if a browser extension is installed or not.
- Previously we were using extension marker DOM element. However, there were 2 issues with it:
   1. It waited up to 3 seconds and then assumed that there is no extension if it loaded later
   2. For self-hosted instances browser extension might not have access to the web and thus not be able to inject any marker element.
- New approach:
  1. Uses the extension marker DOM element or browser extension UTM markers to detect and saves flag (timestamp) in local storage (updates every time when detected)
  2. It uses a local storage flag for up to a week and then only assumes there is no browser extension installed

<details>
<summary>Screenshots </summary>

<img width="828" alt="image" src="https://user-images.githubusercontent.com/6717049/158172923-0987103a-1306-4d21-8548-e76907c5fab7.png">
<img width="654" alt="image" src="https://user-images.githubusercontent.com/6717049/158172987-8a977757-85f6-4d15-87c2-adf0fffc96eb.png">

</details>

## Test plan
1. `sg start web-standalone`
2 Check that Browser extension CTA is shown
   1. Disable browser extension (Right-click > "Manage extension > disable toggle)
   4. Open any search result page or file page: 
       1. for example [search result page](https://sourcegraph.test:3443/search?q=context:global+search&patternType=literal)
       4. for example [file page](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/blob/.eslintrc.js)
   5. It should show up on every mod 6th cadence
1. Check that Browser extension CTA is NOT shown if bext is installed and enabled
   1. Enable browser extension
   5. Open any search result page or file page: 
       1. for example [search result page](https://sourcegraph.test:3443/search?q=context:global+search&patternType=literal)
       3. for example [file page](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/blob/.eslintrc.js)
   6. Check that bext CTA is not shown anymore
1. Check that Browser extension CTA is NOT shown if UTM was detected
   1. Disable browser extension (Right-click > "Manage extension > disable toggle)
   1. Remove `integrations.browser.lastDetectionTimestamp` localStorage item
   4. Open any page with `?utm_source=chrome-extension` parameters
   6. Open any search result page or file page: 
       1. for example [search result page](https://sourcegraph.test:3443/search?q=context:global+search&patternType=literal)
       2. for example [file page](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/blob/.eslintrc.js)
   7. Check that bext CTA is not shown anymore
1. Check that install browser extension popup is not shown on "View in codehost (Github/Gitlab)" button click
   1. Disable browser extension
   1. Remove `has-dismissed-browser-ext-popup` and `integrations.browser.lastDetectionTimestamp` local storage items
   3. Open any [file page](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/blob/.eslintrc.js)
   4. Click on "View on Github" button on the top-right corner
   7. It should show popup to install bext
   8. Enable browser extension and try the same, popup should not show anymore
